batches: implement edit workflow with better page hierarchy

courier-new opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/33395.

I tried to split off as much as this as possible in the previous PRs, but I think we've just got to bite the bullet on the rest of this. Apologies and endless gratitude in advance for any part of this you are able to review. 🙇‍♀️ I swear roughly half of all the lines changed here are from storybook story coverage. But the upside to that is, we *finally* have good storybook coverage on all the major SSBC components! 🙌 🎉

I'd highly recommend giving it a try on the demo app and checking out all the stories under the `batches/batch-spec` namespace before trying to review the code here -- the changes make the most sense when you've already observed the end result. 😅

Biggest decision here is that I opted for React Context over zustand. This is because we have a ton of logic and data fetching closely tied up in the state for the server-side execution workflow, and that's more natural for me to express with Context and pure React hooks. But this could always be refactored to zustand later if that's the direction the frontend is going. The point is that it's nicely packaged up in a hook that all the pages can easily access, not that the hook is implemented with X specific state manager/data provider. 

Note that this also incorporates @adeola-ak's changes from https://github.com/sourcegraph/sourcegraph/pull/34449 (which were already approved and are just waiting on a finalized link to include in for the "Request a demo" button), in order to pre-empt the merge conflicts between them. 🙂

## Test plan

Added like a bajilion and a half stories covering all sorts of states. Tested manually extensively. Logic is intended to be left the same, this just restructures components and reintegrates pieces together differently.
<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

eseliger commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -1506,6 +1506,13 @@ type Query {
         after: String
     ): ExecutorConnection!
 
+    """
+    Returns true if executors have been configured on the Sourcegraph instance.
+    This is based on heuristics and doesn't necessarily mean that they would be
+    working.
+    """
+    areExecutorsConfigured: Boolean!
```
weird, aren't these changes on main?

courier-new replied:
Not yet, they're still unmerged here: https://github.com/sourcegraph/sourcegraph/pull/34449. I think Adeola was waiting to figure out what link to put in the "request a demo" part of the modal. So I just squashed her branch into mine so I could deal with the merge conflicts. 😛

LawnGnome commented on client/web/src/enterprise/batches/batch-spec/edit/workspaces-preview/useWorkspacesPreview.ts:
```diff
@@ -112,6 +111,7 @@ export const useWorkspacesPreview = (
     { isBatchSpecApplied, namespaceID, noCache, onComplete, filters }: UseWorkspacesPreviewOptions
 ): UseWorkspacesPreviewResult => {
     // Track whether the user has previewed the batch spec workspaces at least once.
+    const [hasRequestedPreviewed, setHasRequestedPreviewed] = useState(false)
```
`hasRequestedPreviewed` sounds a bit weird — would this be better phrased `hasRequestedPreview`?

courier-new replied:
Ha, that was actually the intended name. Thanks for flagging that.

LawnGnome commented on client/web/src/enterprise/batches/batch-spec/execute/ReadOnlyBatchSpecForm.tsx:
```diff
@@ -0,0 +1,107 @@
+
+    const alert = batchSpec.isExecuting ? (
+        <Alert variant="warning" className="d-flex align-items-center pr-3">
+            <div className="flex-grow-1">
```
This `div` might need some right padding, looking at the storybook. (Firefox crashed before providing a screenshot.)

courier-new replied:
Good call, thanks!

LawnGnome commented on client/web/src/enterprise/batches/batch-spec/execute/ReadOnlyBatchSpecForm.tsx:
```diff
@@ -0,0 +1,107 @@
+        </Alert>
+    ) : (
+        <Alert variant="info" className="d-flex align-items-center pr-3">
+            <div className="flex-grow-1">
```
Same issue here.

LawnGnome commented on client/web/src/enterprise/batches/create/ConfigurationForm.tsx:
```diff
@@ -0,0 +1,211 @@
+                        Give it a short, descriptive name to reference the batch change on Sourcegraph. Do not include
+                        confidential information.{' '}
+                        <span className={classNames(isNameValid === false && 'text-danger')}>
+                            Only regular characters, _ and - are allowed.
```
Sorry, I know you didn't add this in this PR, but:

```suggestion
                            Only letters, numbers, _ and - are allowed.
```

LawnGnome commented on client/web/src/enterprise/batches/create/InsightTemplatesBanner.tsx:
```diff
@@ -14,7 +14,7 @@ interface InsightTemplatesBannerProps {
 }
 
-export const InsightTemplatesBanner: React.FunctionComponent<InsightTemplatesBannerProps> = ({
+export const InsightTemplatesBanner: React.FunctionComponent<React.PropsWithChildren<InsightTemplatesBannerProps>> = ({
```
Do we use the children anywhere?

LawnGnome reviewed (approved):
Not a super thorough review, but LGTM, and I looked at all the storybooks. Good job!

eseliger reviewed (approved):
Really, really good improvements here, it feels much easier to navigate right away :) 

Wrote down some notes during manual testing:

did this somehow affect the styles for this section? it looks a bit busy and I think it used to be different:
<img width="394" alt="Screenshot 2022-05-17 at 17 03 26@2x" src="https://user-images.githubusercontent.com/19534377/168857124-a6e738be-1913-4abc-beb2-c61916c91582.png">

Should canceling a workflow get you right back to the edit page? Once back there, you're not allowed to proceed to the edit page anymore and hence you lose access to all logs and errors immediately.

Is this based on latest main? The show more is kil again 😬 

Is it intentional that the edit spec button opens a new tab?

courier-new commented:
> did this somehow affect the styles for this section? it looks a bit busy and I think it used to be different

Good call, it breaks onto a new line when the text changes from "add" (before you've previewed) to "modify" (after you've previewed). I don't even think we need the example anymore after you've previewed, so I just removed the button and added a storybook to cover that state. 

> Should canceling a workflow get you right back to the edit page? Once back there, you're not allowed to proceed to the edit page anymore and hence you lose access to all logs and errors immediately.

Hmm, yeah, I suppose that makes sense in the case that you canceled from "edit" but not if you just purely canceled. I'll update that.

> Is it intentional that the edit spec button opens a new tab?

Yeah, this was intended to help with the iteration aspect of editing: you can easily reference what you started with on one tab while you work on modifying it in another. I do wonder if it's confusing if you forget which tab is the more up-to-date one, though. Since previewing the workspaces actually saves a new state of the batch spec as the latest, it could cause user error if they accidentally overwrite their batch spec with an older version of it. But at least the tab they opened it on has the read-only banner?

> Is this based on latest main? The show more is kil again

Not to worry, it's out of date as of yesterday!!

eseliger commented:
> Yeah, this was intended to help with the iteration aspect of editing: you can easily reference what you started with on one tab while you work on modifying it in another. I do wonder if it's confusing if you forget which tab is the more up-to-date one, though. Since previewing the workspaces actually saves a new state of the batch spec as the latest, it could cause user error if they accidentally overwrite their batch spec with an older version of it. But at least the tab they opened it on has the read-only banner?

As you start a new version of an unapplied batch spec (ie. you hit edit), the old batch spec is replaced by a new one and eventually, the other tab would show the new batch spec's content, I think. Is this still helpful in that case? Just wondering because it stays read only in that tab and becomes editable in the other.

courier-new commented:
> eventually, the other tab would show the new batch spec's content, I think. Is this still helpful in that case?

I think it would only show the new one in the old tab if some event triggered a refetch of the batch spec, like navigating away from the editor and back again or refreshing the tab. AFAIK just refocusing the tab doesn't trigger refetch, so it wouldn't happen until the user took some deliberate action on the page. But yes, if that did happen, the two tabs would not be useful anymore.

Perhaps since we haven't explicitly made much design consideration over a multi-tab workflow before, I'll nix the open-in-new-tab behavior and leave a comment to explore it in the future.

courier-new commented on client/web/src/enterprise/batches/preview/BatchChangePreviewPage.tsx:
```diff
@@ -137,14 +141,14 @@ export const NewBatchChangePreviewPage: React.FunctionComponent<
         )
     }
-    if (data?.node?.__typename !== 'BatchSpec') {
+    if (data?.node?.__typename !== 'BatchSpec' || error) {
         return <HeroPage icon={AlertCircleIcon} title="Batch spec not found" />
     }
```
Yep, thanks, I think I copied this from somewhere else unintentionally -- I'll update in both places.
